landoapi/hg.py

```diff
@@ -368,56 +372,158 @@ def apply_patch(self, patch_io_buf):
+        parser.read_string(lando_ini_contents)
 
-        # If the file doesn't have a `fix` section, exit early.
-        if not parser.has_section("fix"):
-            return None
+        return parser
 
-        fix_hg_command = []
-        for key, value in parser.items("fix"):
-            if not key.endswith(":pattern"):
-                continue
+    def run_code_formatters(self) -> str:
+        """Run automated code formatters, returning the output of the process.
 
-            fix_hg_command += ["--config", f"fix.{key}={value}"]
+        Changes made by code formatters are applied to the working directory and
+        are not committed into version control.
+        """
+        return self.run_mach_command(["lint", "--fix", "--outgoing"])
 
-        # Exit if we didn't find any patterns.
-        if not fix_hg_command:
-            return None
+    def run_mach_bootstrap(self) -> str:
+        """Run `mach bootstrap` to configure the system for code formatting."""
+        return self.run_mach_command(
+            [
+                "bootstrap",
+                "--no-system-changes",
+                "--application-choice",
+                "browser",
+            ]
+        )
 
-        # Run the formatters.
-        fix_hg_command += ["fix", "-r", "stack()"]
-        fix_output = self.run_hg(fix_hg_command).splitlines()
+    def run_mach_command(self, args: List[str]) -> str:
+        """Run a command using the local `mach`, raising if it is missing."""
+        if not self.mach_path:
+            raise Exception("No `mach` found in local repo!")
+
+        # Convert to `str` here so we can log the mach path.
+        command_args = [str(self.mach_path)] + args
+
+        try:
+            logger.info("running mach command", extra={"command": command_args})
+
+            output = subprocess.run(
+                command_args,
+                capture_output=True,
+                check=True,
+                cwd=self.path,
+                encoding="utf-8",
+                universal_newlines=True,
+            )
+
+            logger.info(
+                "output from mach command",
+                extra={
+                    "output": output.stdout,
+                },
+            )
+
+            return output.stdout
+
+        except subprocess.CalledProcessError as exc:
+            logger.exception(
+                "Failed to run mach command",
+                extra={
+                    "command": command_args,
+                    "err": exc.stderr,
+                    "output": exc.stdout,
+                },
+            )
 
-        # Update the working directory to the latest change.
-        self.run_hg(["update", "-C", "-r", "tip"])
+            raise exc
 
-        # Exit if no revisions were reformatted.
-        pre_formatting_hashes = check_fix_output_for_replacements(fix_output)
-        if not pre_formatting_hashes:
+    def format_stack_amend(self) -> Optional[List[str]]:
```

Comment:
Not very certain why we should treat 1-revision stacks any differently than multi-revision stacks.

Reply:
Since we know formatting changes made to 1-revision stacks all apply to the revision, we can just amend them in. This avoids creating any autoformatting commits for the most common type of landings.